src/core/probe/builder.rs

```diff
@@ -37,6 +49,71 @@ pub(super) fn reuse_map_fds(
     Ok(())
 }
 
+/// Replaces raw bpf programs using both btf-rs and bpf_sys
+/// facilities. This is currently needed because libbpf-rs doesn't
+/// have loading nor replacing capabilities for non-ELF files.
+fn replace_raw_filter(filter: &Filter, fd: u32, tgt_func: &str) -> Result<()> {
+    let btf_buff = bpf_sys::get_btf_from_fd(fd)?;
+    let btf = Btf::from_bytes(&btf_buff)?;
+    let tgt_btf_id = btf.resolve_id_by_name(tgt_func)?;
+
+    match btf.resolve_type_by_id(tgt_btf_id)? {
+        Type::Func(_) => (),
+        _ => bail!("Resolved type is not a function"),
+    }
+
+    let btf_fd = bpf_sys::load_btf(crate::core::filters::packets::filter_stub::BTF)?;
+
+    let btf = Btf::from_bytes(crate::core::filters::packets::filter_stub::BTF)?;
+    let (func_info_rec, rec_size, rec_count) =
+        bpf_sys::gen_dummy_func_info_rec(btf.resolve_id_by_name(tgt_func)?);
+
+    let mut attrs: bpf_sys::bpf_attr = unsafe { std::mem::zeroed() };
+    let load_attrs = unsafe { &mut attrs.__bindgen_anon_3 };
+
+    load_attrs.prog_type = bpf_sys::bpf_prog_type::BPF_PROG_TYPE_EXT as u32;
+    let prog_name =
+        CString::new(format!("r_{}", tgt_func)).expect("new string for prog name failed");
+    unsafe {
+        ptr::copy_nonoverlapping(
+            prog_name.as_ptr(),
+            load_attrs.prog_name.as_mut_ptr(),
+            load_attrs.prog_name.len(),
+        )
+    };
+
+    load_attrs.insn_cnt = (filter.prog.len() / 8) as u32;
+    let prog: &[u8] = &filter.prog;
+    load_attrs.insns = prog.as_ptr() as u64;
+    let license = CString::new("GPL").expect("new string for license failed");
+    load_attrs.license = license.as_ptr() as u64;
+    load_attrs.attach_btf_id = tgt_btf_id;
+    load_attrs.__bindgen_anon_1.attach_prog_fd = fd;
+    load_attrs.prog_btf_fd = btf_fd;
+    load_attrs.func_info = &func_info_rec as *const _ as u64;
+    load_attrs.func_info_rec_size = rec_size as u32;
+    load_attrs.func_info_cnt = rec_count;
+
+    let load_fd = bpf_sys::bpf(bpf_sys::bpf_cmd::BPF_PROG_LOAD, &attrs)?;
+    let mut attrs: bpf_sys::bpf_attr = unsafe { std::mem::zeroed() };
+    let link_attr = unsafe { &mut attrs.link_create };
+    link_attr.prog_fd = load_fd;
+
+    bpf_sys::bpf(bpf_sys::bpf_cmd::BPF_LINK_CREATE, &attrs)?;
+
+    Ok(())
+}
+
+pub(super) fn replace_filter(fd: i32, filters: &[FilterType]) -> Result<()> {
```

Comment:
not really now, but I guess it will relatively soon.
There should be no way to provide multiple filters of the same type, though, but a check will increase the robustness here.
Technically, this is mostly speculation, although we could end up having multiple filters (there's no infra bpf side for it, though), only one per type should be present. I don't foresee something like having two "Packet" filters per probe. I guess if we find ourselves in such a situation, we should manage it somewhere else, as long as it is possible.